### PR TITLE
Refactor direct fixture call warning to avoid incompatibility with plugins

### LINKED ISSUE #3720: Improve lookup/request to allow access to class instances from staticfunctions
Follow up from https://github.com/pytest-dev/pytest/pull/3705#discussion_r204254468

We would like to cleanup the `__being_called_by_pytest` keyword hack that was used in #3705. @RonnyPfannschmidt's suggestion of unwrapping the fixture function just before the call works great, except that for staticmethods we don't have access to instance request at that point.

It seems like a deeper issue in the structural construction of the request: static test functions should still have `self` in `request.instance`.

### LINKED ISSUE #3747: 3.7.0: Fixture misses attribute
From https://github.com/pytest-dev/pytest/issues/3742#issuecomment-409170661: 

Hello! Thank you for this great library! The Mozilla release engineering team relies on it for several projects :smiley: 

I just found a similar regression:
``` py
AttributeError: 'async_generator' object has no attribute 'put'
```
This async generator is a regular `aiohttp.ClientSession()` and is [initialized here](https://github.com/mozilla-releng/scriptworker/blob/cbd8ba44aa8ed1d3eeaf40d7d2a987f887074171/scriptworker/test/__init__.py#L180). This might be an integration issue with the `pytest-asyncio` plugin. 

I rolled pytest back to 3.6.4 and things are back working. You can see both the failure and the workaround in

* https://github.com/mozilla-releng/beetmoverscript/pull/176
* https://travis-ci.org/mozilla-releng/beetmoverscript/jobs/410250756#L598

## Example

I put up a minimal example over here: https://github.com/JohanLorenzo/pytest-fixture-bug-3747

### PR DESCRIPTION
**Pull request by nicoddemus**
This refactors the code so we have the real function object right during
collection. This avoids having to unwrap it later and lose attached information
such as "async" functions.

I've tested this with `pytest-asyncio` and it fixes https://github.com/pytest-dev/pytest-asyncio/issues/89.

Fix #3747, Fix #3720
Fix https://github.com/pytest-dev/pytest-asyncio/issues/89

### CODE REVIEW

File: changelog/3747.bugfix.rst [hunk 1]
```diff
@@ -0,0 +1 @@
+Fix compatibility problem with plugins and the the warning code issued by fixture functions when they are called directly.
```
Comment (Callek) [inline changelog/3747.bugfix.rst]: drive-by-nit: "the the"

Comment (nicoddemus) [reply changelog/3747.bugfix.rst]: Thanks!

**Comment by nicoddemus** [comment]
Hmmm 2 failing tests on py27, I will take a look at them tomorrow if the current approach is considered sound by others.

File: src/_pytest/compat.py [hunk 2]
```diff
@@ -249,6 +249,21 @@ def get_real_func(obj):
     return obj
 
 
+def get_real_method(obj, holder):
```
Comment (RonnyPfannschmidt) [inline src/_pytest/compat.py]: this needs to handle `__wrapped__` on python2

Comment (nicoddemus) [reply src/_pytest/compat.py]: Hmm `get_real_func` already handles `__wrapped__`. Can you clarify this a bit?

Comment (RonnyPfannschmidt) [reply src/_pytest/compat.py]: @nicoddemus my understanding is that method descriptors are not exposing it readily, i will investigate in a bit

File: src/_pytest/fixtures.py [hunk 3]
```diff
@@ -1303,6 +1293,15 @@ def parsefactories(self, node_or_obj, nodeid=NOTSET, unittest=False):
                     name = marker.name
                 assert not name.startswith(self._argprefix), FIXTURE_MSG.format(name)
 
+            # during fixture definition we wrap the original fixture function
+            # to issue a warning if called directly, so here we unwrap it in order to not emit the warning
+            # when pytest itself calls the fixture function
+            if six.PY2 and unittest:
```
Comment (nicoddemus) [inline src/_pytest/fixtures.py]: This was the best I could come with; unbound methods are a headache to deal here.

Comment (RonnyPfannschmidt) [reply src/_pytest/fixtures.py]: cant wait for the python2 drop ^^

Comment (nicoddemus) [reply src/_pytest/fixtures.py]: You and me both! :grin:
